{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "897f50f7",
   "metadata": {},
   "source": [
    "# Exception"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c8d6dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The error that occurred at runtime is know as an exception.\n",
    "# The mechanism that handles error at runtime is called exception handling.\n",
    "\n",
    "# The BaseException is the parent/root class of all exceptions classes.\n",
    "# It is not directly recommended that catching BaseException bcz it includes\n",
    "# system-exiting exceptions like SystemExit, KeyboardInterrupt, and GeneratorExit, among others.\n",
    "\n",
    "# Instead, Python programmers typically catch more specific exceptions or subclass Exception,\n",
    "# which is a more common base class for user-defined exceptions and application-specific error handling.\n",
    "# Exception itself is a subclass of BaseException, so it inherits its properties while being more\n",
    "# appropriate for general error handling in Python code.\n",
    "\n",
    "# In python exceptions are generated by the python interpreter\n",
    "# In python there is no checked and unchecked exceptions.\n",
    "\n",
    "# try\n",
    "# The try is the block of code that contain the code that may raise an exception.\n",
    "# The exception throw from try block is caught by the except block.\n",
    "# For every try block there can be one or more than one except block or a finally block\n",
    "\n",
    "# except\n",
    "# The except is the block of code that contain the code that will handle the exception.\n",
    "\n",
    "# finally\n",
    "# The finally block is the always excutable block either the exception occurred or not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "110366e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here a Zero Division Error Occurred\n",
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "print(\"Here a Zero Division Error Occurred\")\n",
    "try:\n",
    "  c = 10 / 0\n",
    "  print(c)\n",
    "# Here we use Exception class as a base class for all exceptions.\n",
    "# If the BaseException class is used then it will catch all the exception like system exception\n",
    "except Exception as e:\n",
    "  print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b73edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try Block: The code inside the try block is executed. In this case, 10 / 0 causes a ZeroDivisionError because division by zero is not allowed in Python.\n",
    "\n",
    "# Exception Raised: When the ZeroDivisionError occurs, Python creates an exception object representing the error. This object contains information about the error, including a message that describes the error (\"division by zero\" in this case).\n",
    "\n",
    "# except ZeroDivisionError as e: The except block catches the ZeroDivisionError. The as e part assigns the exception object to the variable e.\n",
    "\n",
    "# Accessing the Exception Object: Inside the except block, you can use the variable e to access the exception object. When you print e, it calls the __str__ method of the exception object, which returns the error message.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26abe0e4",
   "metadata": {},
   "source": [
    "# Types of Errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "80c52527",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------ZeroDivisionError-------------------------------\n",
      "Here 10 is divided by zero\n",
      "So the error is division by zero\n",
      "Bye\n"
     ]
    }
   ],
   "source": [
    "print(\"---------------------------ZeroDivisionError-------------------------------\")\n",
    "# The ZeroDivisionError is raised when the second argument of a division or modulo operation is zero.\n",
    "try:\n",
    "  print(10 / 0)\n",
    "except ZeroDivisionError as e:\n",
    "  print(\"Here 10 is divided by zero\")\n",
    "  print(\"So the error is\", e)\n",
    "print(\"Bye\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cfddfd8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------TypedError---------------------------------\n",
      "The string hello can't be added with integer 10 so this will be an error\n",
      "The error is can only concatenate str (not \"int\") to str\n",
      "Bye\n"
     ]
    }
   ],
   "source": [
    "print(\"----------------------TypedError---------------------------------\")\n",
    "try:\n",
    "  c = \"Hello\" + 10\n",
    "except TypeError as e:\n",
    "  print(\"The string hello can't be added with integer 10 so this will be an error\")\n",
    "  print(\"The error is\", e)\n",
    "print(\"Bye\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "18a2cd9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------ValueError---------------------------\n",
      "This is a type of ValueError bcz the value can't be converted to integer\n",
      "Bye\n"
     ]
    }
   ],
   "source": [
    "print(\"-------------------------ValueError---------------------------\")\n",
    "try:\n",
    "  c = int(\"abc\")\n",
    "  print(c)\n",
    "except ValueError as e:\n",
    "  print(\"This is a type of ValueError bcz the value can't be converted to integer\")\n",
    "print(\"Bye\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0aecbf55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------IndexError---------------------------------\n",
      "We are trying to accessing the index 5 which is not present in the list\n",
      "The error is list index out of range\n",
      "Bye\n"
     ]
    }
   ],
   "source": [
    "print(\"-------------------------IndexError---------------------------------\")\n",
    "try:\n",
    "  l1 = [1, 2, 3, 4, 5]\n",
    "  print(l1[5])\n",
    "except Exception as e:\n",
    "  print(\"We are trying to accessing the index 5 which is not present in the list\")\n",
    "  print(\"The error is\", e)\n",
    "print(\"Bye\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c578101",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------KeyError----------------------------------\n",
      "Here we are trying to access the value of the age which is not present in the dictionary\n",
      "So in the place of the error the key will be printed for which we are trying to access the value 'age'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"------------------------KeyError----------------------------------\")\n",
    "try:\n",
    "  d1 = {\"name\": \"Gogo\"}\n",
    "  print(d1[\"age\"])\n",
    "except KeyError as e:\n",
    "  print(\"Here we are trying to access the value of the age which is not present in the dictionary\")\n",
    "  print(\"So in the place of the error the key will be printed for which we are trying to access the value\",e)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3a14d1ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The c is not defined and we are trying to print it\n",
      "So the error is  name 'c' is not defined\n",
      "Bye\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "  print(c)\n",
    "except NameError as e:\n",
    "  print(\"The c is not defined and we are trying to print it\")\n",
    "  print(\"So the error is \", e)\n",
    "print(\"Bye\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0381797",
   "metadata": {},
   "source": [
    "# Single Try Mutiple Except"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d28fb1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n",
      "<class 'NameError'>\n"
     ]
    }
   ],
   "source": [
    "l1 = [10, 20, 30, 40]\n",
    "# Here there are two error in the code\n",
    "# Whatever is the error which come first it will be executed\n",
    "# the zeroDivision error is coming first\n",
    "try:\n",
    "  print(10 / 0)\n",
    "  print(l1[5])\n",
    "except ZeroDivisionError as e:\n",
    "  print(e)\n",
    "except IndexError:\n",
    "  print(\"Index Error\")\n",
    "\n",
    "try:\n",
    "  print(x)\n",
    "except NameError as e:\n",
    "  print(type(e))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e4f21b2",
   "metadata": {},
   "source": [
    "# Nested Try and Except"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad9a0522",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n",
      "NameError\n",
      "name 'd' is not defined\n",
      "SomeError\n"
     ]
    }
   ],
   "source": [
    "a = 10\n",
    "b = 0\n",
    "try:\n",
    "  try:\n",
    "    c = a / 0\n",
    "  except ZeroDivisionError as e:\n",
    "    print(e)\n",
    "  try:\n",
    "    print(d)\n",
    "  except:\n",
    "    print(\"NameError\")\n",
    "except:\n",
    "  print(\"SomeError\")\n",
    "\n",
    "try:\n",
    "  print(d)\n",
    "  # Here a message is printed\n",
    "except NameError as e:\n",
    "  print(e)\n",
    "\n",
    "try:\n",
    "  print(e)\n",
    "except:\n",
    "  print(\"SomeError\")\n",
    "\n",
    "# Here we are not catching any specify exception what ever the exception comes we are going to printing the message\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49241e2",
   "metadata": {},
   "source": [
    "# finally block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6ba2c0c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It is a always excutable block\n",
      "0.5\n",
      "Some Error\n",
      "It is a always excutable block\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# The finally block is always executable block either the exception occurred or not\n",
    "\n",
    "# The finally block is usuaully used with return statement bcz if we return from statement and we want to print the a particular statement  every time then we use finally block\n",
    "\n",
    "\n",
    "def div(a, b):\n",
    "  try:\n",
    "    c = a / b\n",
    "    return c\n",
    "  except:\n",
    "    print(\"Some Error\")\n",
    "  finally:\n",
    "    print(\"It is a always excutable block\")\n",
    "\n",
    "\n",
    "res = div(10, 20)\n",
    "print(res)\n",
    "\n",
    "res = div(10, 0)\n",
    "print(res)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403d1b3b",
   "metadata": {},
   "source": [
    "# Raise Keyword"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b62fea48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In Python, the raise statement is used to explicitly raise an exception or error during program execution. When you use the raise statement, you indicate to Python that a particular condition or situation warrants an exception, and you specify the type of exception to be raised.\n",
    "\n",
    "# The raise statement is used to raise an exception, which is an error that occurs during program execution.\n",
    "\n",
    "# The general syntax of the raise statement is as follows:\n",
    "\n",
    "# raise exception_type(\"Optional error message\")\n",
    "# Here's a breakdown of each part of the raise statement:\n",
    "\n",
    "# raise: This keyword is used to raise an exception.\n",
    "\n",
    "# exception_type: This specifies the type of exception to be raised. It can be a built-in exception class\n",
    "# (e.g., ValueError, TypeError, ZeroDivisionError, etc.) or a custom exception class that you have defined.\n",
    "\n",
    "# \"Optional error message\": This is an optional argument that allows you to provide additional information about the exception.\n",
    "# It is typically a string describing the reason for raising the exception."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5e02c6bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raising an own exception divide by one using raise keyword\n",
      "Divide by one\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Raising an own exception divide by one using raise keyword\")\n",
    "a = 10\n",
    "b = 1\n",
    "try:\n",
    "  if (b == 1):\n",
    "    raise ZeroDivisionError(\"Divide by one\")\n",
    "  else:\n",
    "    print(a / b)\n",
    "except ZeroDivisionError as e:\n",
    "  print(e)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0cfc6c74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here, ZeroDivisionError(\"Divide by one\") creates an instance of the ZeroDivisionError class with the message Divide by one\"\n",
    "# This object is then thrown by Python.\n",
    "\n",
    "# On writing ZeroDivisionError(\"Divide by one\") the __init__ method of the ZeroDivisionError class.\n",
    "\n",
    "# The message \"Divide by zero\" is passed to the constructor (__init__ method) of the ZeroDivisionError class.\n",
    "\n",
    "# The base Exception class stores this message in the args attribute of the instance as a tuple\n",
    "# like that\n",
    "\n",
    "# def __init__(self, *args):\n",
    "# self.args = *args\n",
    "\n",
    "# And on printing e the message division by zero is printed bcz on printing object python call the __str__ method of the object and that method returns the message.(self.args[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ace6be11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating user define exception\n",
      "This is the exception of the MyTest class which ocurrs on dividing the number by 1\n"
     ]
    }
   ],
   "source": [
    "print(\"Creating user define exception\")\n",
    "# First we need to create class and our class must be the child class of the Exception class.\n",
    "\n",
    "\n",
    "class MyTest(Exception):\n",
    "\n",
    "  def __init__(self, a, b):\n",
    "    self.a = a\n",
    "    self.b = b\n",
    "\n",
    "  def div(self):\n",
    "    try:\n",
    "      if (self.b == 1):\n",
    "        raise MyTest(a, b)\n",
    "      else:\n",
    "        return self.a / self.b\n",
    "    except AttributeError:\n",
    "      print(\"Some Error\")\n",
    "\n",
    "  def __str__(self):\n",
    "    return \"This is the exception of the MyTest class which ocurrs on dividing the number by 1\"\n",
    "\n",
    "\n",
    "t1 = MyTest(10, 1)\n",
    "try:\n",
    "  t1.div()\n",
    "except MyTest as e:\n",
    "  print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa2f8d5",
   "metadata": {},
   "source": [
    "# if __name__==''__main__''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9c65b8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In python if we import an module then we can access the functions and variables defined in that module\n",
    "# But if the imported module also print something then default that is also printed in our code also\n",
    "\n",
    "# Example\n",
    "# Here I have imported the module named mymodule and then I have used the function\n",
    "# defined in that module\n",
    "# The module that i am importing is mymodule.py also print the function due to which\n",
    "# a definition of the function is printed ones extra then it is called.\n",
    "\n",
    "# import mymodule as my\n",
    "\n",
    "# my.welcome()\n",
    "\n",
    "# Here the definition of the function welcome is printed twice bcz the module that we are importing is also calling welcome function due to which we getting the output twice.\n",
    "\n",
    "# To avoid that in python we have if __name__==\"__main__\"\n",
    "\n",
    "# In Python, the if __name__ == \"__main__\": block is used to define the main entry point of a Python script or module. It allows you to write code that will only be executed when the script is run directly, not when it is imported as a module into another script.\n",
    "'''\n",
    "\n",
    "Here's a breakdown of how it works:\n",
    "\n",
    "Module Execution vs. Import:\n",
    "When a Python script is executed directly (e.g., using python script.py), it is considered the main module being run. On the other hand, if the script is imported as a module into another script, it is not the main module but rather a module being imported.\n",
    "\n",
    "__name__ Attribute:\n",
    "Every Python module has a built-in __name__ attribute. When the module is run directly, __name__ is set to \"__main__\". When the module is imported, __name__ is set to the module's name (e.g., \"module_name\").\n",
    "\n",
    "Usage of if __name__ == \"__main__\": block:\n",
    "By using if __name__ == \"__main__\":, you can write code that should only run when the script is executed directly. This is commonly used for script initialization, testing, or running specific tasks when the script is invoked directly from the command line.\n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdee29a0",
   "metadata": {},
   "source": [
    "# Module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "047bfb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec21548c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.1622776601683795\n"
     ]
    }
   ],
   "source": [
    "print(math.sqrt(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f3dc9db9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0\n"
     ]
    }
   ],
   "source": [
    "print(math.sqrt(4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf1b928b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import sqrt,pi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f59ccd63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.23606797749979"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sqrt(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f91f2bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.141592653589793"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3433ab72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.5440211108893698"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "math.sin(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0db398c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0adc5ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.5440211108893698"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sin(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9208a2b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sqrt(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6b848ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.1622776601683795"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "math.sqrt(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "991dcdbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sqrt(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "65279bf0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6fa5f5a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is my first module\n"
     ]
    }
   ],
   "source": [
    "# Here we are importing the our own created module named as mymodule.\n",
    "import mymodule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5cdde4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# On again importing the module we will not get the data.\n",
    "# We avoid that we use module name as imp.\n",
    "# In which we have use reload method to reload the module to get its content.\n",
    "import mymodule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1704648b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\n\\nWhen you import a module in Jupyter Notebook (or in Python in general), the module is loaded and executed only once per session. \\nSubsequent imports of the same module do not re-execute the module's code. \\nThis behavior is due to Python's internal module caching mechanism, which stores already-loaded modules in a dictionary called sys.modules.\\n\\nIf you attempt to re-import the same module during the same session, Python retrieves it from this cache rather than re-executing its code.\\n\""
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Resaon \n",
    "'''\n",
    "\n",
    "When you import a module in Jupyter Notebook (or in Python in general), the module is loaded and executed only once per session. \n",
    "Subsequent imports of the same module do not re-execute the module's code. \n",
    "This behavior is due to Python's internal module caching mechanism, which stores already-loaded modules in a dictionary called sys.modules.\n",
    "\n",
    "If you attempt to re-import the same module during the same session, Python retrieves it from this cache rather than re-executing its code.\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b4add95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mymodule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "389bce5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mymodule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1a4f5e3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import imp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a5b8b7b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is my first module\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<module 'mymodule' from 'C:\\\\Users\\\\Aryan\\\\mymodule.py'>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imp.reload(mymodule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "211e9346",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mymodule1 as my"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "591a45aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my.add(10,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6ad187c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "200"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my.multiply(10,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cc749f46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100000000000000000000"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my.power(10,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "70f90438",
   "metadata": {},
   "outputs": [],
   "source": [
    "import test6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6d583595",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'GOGO',\n",
       " 'courses': ['ml', 'dl', 'cv', 'stats'],\n",
       " 'msg': 'List of all courses'}"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "28f2f388",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'List of all courses'"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.msg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7cc46b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "import test6\n",
    "import imp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "794a8b10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'GOGO',\n",
       " 'courses': ['ml', 'dl', 'cv', 'stats'],\n",
       " 'msg': 'List of all courses'}"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "0d6bb6de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'test6' from 'C:\\\\Users\\\\Aryan\\\\test6.py'>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imp.reload(test6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7d3980aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'GOGO',\n",
       " 'course': ['ml', 'dl', 'cv', 'stats'],\n",
       " 'msg': 'List of all courses'}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "6427eebe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ml', 'dl', 'cv', 'stats']"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.get_courses()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ced70017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'List of all courses'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test6.msg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "12d88552",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66527924",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.getcwd()+\"\\test2\"):\n",
    "    os.mkdir(\"test2\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "8d9e24d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "fde556c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('C:/Users/Aryan/test2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "2afea7c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan\\\\test2'"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "0ff17330",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.getcwd()+'\\test'):\n",
    "    os.mkdir('test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "50d2116c",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(os.getcwd()+'\\\\test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6444b38c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan\\\\test2\\\\test'"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "310d88dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1,4):\n",
    "    f = open(f'mod{i}.py','w')\n",
    "    function_code = f'''def fn{i}():\n",
    "    return \"This is a fn1 function\"\n",
    "    '''\n",
    "    f = open(f\"mod{i}.py\",'w')\n",
    "    f.write(function_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "90a32a83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan\\\\test2\\\\test'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "1d46e2ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan\\\\test2\\\\test'"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "dfb56bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "funtion_code = '''\n",
    "def fn1():\n",
    "    return \"This is a fn1 function\"\n",
    "'''\n",
    "f = open(\"mod1.py\",'w')\n",
    "f.write(function_code)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "4e7af545",
   "metadata": {},
   "outputs": [],
   "source": [
    "from test2.test123 import mod1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "9bba2ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'This is function11'"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod1.fn11()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "fffb75ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Aryan\\\\test2\\\\test'"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "8b40b176",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('C:\\\\Users\\\\Aryan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "fdd7b199",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "__init__.py\n"
     ]
    }
   ],
   "source": [
    "# Including an empty __init__.py file in a directory was traditionally required to treat the directory as a package in Python. Here's why it was important and the current state of its necessity:\n",
    "\n",
    "# Why Was __init__.py Important?\n",
    "\n",
    "# 1.Package Recognition in Python 2:\n",
    "# In Python 2, a directory without an __init__.py file was not treated as a package.\n",
    "# Including the file indicated to the Python interpreter that the directory was a Python package and could be imported.\n",
    "\n",
    "# 2.Initialization and Configuration:\n",
    "\n",
    "# The __init__.py file could contain initialization code for the package.\n",
    "# For example, you could:\n",
    "# Import specific submodules or symbols to expose them at the package level.\n",
    "# Set up package-level variables or configurations.\n",
    "\n",
    "# 3.Explicit Declaration:\n",
    "\n",
    "# It served as an explicit declaration that a directory was intended to be a Python package, reducing accidental misinterpretation.\n",
    "print(\"__init__.py\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe64948",
   "metadata": {},
   "source": [
    "# Compile Time Errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "c91374c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python, being an interpreted language, primarily encounters errors at runtime rather than compile time. \n",
    "# However, there are still compile-time errors that can occur when Python attempts to convert your code into bytecode before executing it. \n",
    "# These errors usually involve syntax errors or issues detected during parsing. Here's an overview:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37cd54f2",
   "metadata": {},
   "source": [
    "# What Are Compile-Time Errors in Python?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "dfe6da8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.Syntax Errors:\n",
    "# These occur when the Python parser detects invalid syntax in your code\n",
    "\n",
    "# Here the double quote of the string is not closed.\n",
    "# So it is the syntax error which unable to code to convert it into bypte code.\n",
    "# print(\"fjndsfkjnfksd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "cf274257",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.Indentation Errors:\n",
    "# Python relies on proper indentation for defining code blocks. Improper indentation results in a compile-time error.\n",
    "# def fun():\n",
    "# print(\"THis is function\")\n",
    "\n",
    "# Here in the code print statment must be written after 4 tabs space to indicate that it is the part of the function.\n",
    "# Since it is not so, it gives error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "c2a625ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.Name Errors at Parse Time:\n",
    "# If you reference a keyword incorrectly or attempt to use an invalid identifier name, it results in a syntax error.\n",
    "\n",
    "\n",
    "# Here in the code a variable name is used from the reversed keywords of the python so it gives error.\n",
    "# class = 10  # \"class\" is a reserved keyword\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e99272a",
   "metadata": {},
   "source": [
    "# What is Error "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "b26a9cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# An error refers to a problem in the program that causes it to terminate abnormally. \n",
    "# Errors often result from programming mistakes or unforeseen conditions, \n",
    "# and they are classified into two main types:\n",
    "\n",
    "# 1. Compile time error.\n",
    "# 2. Run time error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "f494cb36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Syntax Errors\n",
    "# Occur when the Python interpreter cannot parse your code because it violates the language's syntax rules.\n",
    "# These are detected at compile time and must be fixed for the code to run.\n",
    "\n",
    "\n",
    "\n",
    "# Compilation in Python\n",
    "# When you run a Python script (.py file), Python first compiles the code into bytecode (an intermediate format).\n",
    "# This bytecode is stored as .pyc files (Python compiled files) inside the __pycache__ directory.\n",
    "# This compilation step is automatic and not visible to the user.\n",
    "# Example:\n",
    "\n",
    "# print(\"Hello, name !\")\n",
    "# Python first compiles it into bytecode before execution.\n",
    "\n",
    "\n",
    "# Interpretation in Python\n",
    "# The Python Interpreter (CPython, PyPy, etc.) takes the bytecode and executes it line-by-line.\n",
    "# This is why Python is considered an interpreted language—it does not require explicit compilation like C or Java.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "04666a5a",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "expected ':' (1863056084.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  Cell \u001b[1;32mIn[145], line 1\u001b[1;36m\u001b[0m\n\u001b[1;33m    if True\u001b[0m\n\u001b[1;37m           ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m expected ':'\n"
     ]
    }
   ],
   "source": [
    "if True\n",
    "print(\"yes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "fb241ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.Runtime Errors\n",
    "# Occur during the execution of a program, even if the code is syntactically correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "ed2a4994",
   "metadata": {},
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[144], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;241m10\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m0\u001b[39m)\n",
      "\u001b[1;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "print(10/0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "5a2776ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "# This code will catch all the exception as the specific exception class is not mentioned after except block.\n",
    "# It is important to write the exception class after the except block bcz it makes debugging easier as we dont which type of exception can occur at the runtime. \n",
    "try:\n",
    "    print(10/0)\n",
    "except:\n",
    "    print(\"This is a error\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "d28055e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(10/0)\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "472d68aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing the except block and the error is division by zero\n"
     ]
    }
   ],
   "source": [
    "# In python else statment can be used with try and except block also.\n",
    "# else statement which is written just after except block only be executed if the except block is not executed.\n",
    "\n",
    "try:\n",
    "    print(10/0)\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"Executing the except block and the error is\",e)\n",
    "else:\n",
    "    print(\"The successfull execution of the else statement\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "d3400b9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the first number : 10\n",
      "Enter the second number : 20\n",
      "The else statement after the except block\n",
      "line3\n",
      "line2line4\n",
      "line5\n",
      "line6\n",
      "line7\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    a = int(input(\"Enter the first number : \"))\n",
    "    b = int(input(\"Enter the second number : \"))\n",
    "    f = open(\"test1.txt\",'r')\n",
    "    f.read()\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"There is the issue with code\",e)\n",
    "    f = open('test1.txt','r')\n",
    "    print(f.read())\n",
    "else:\n",
    "    print(\"The else statement after the except block\")\n",
    "    try:\n",
    "        f = open('test3.txt','r')\n",
    "        print(f.read())\n",
    "    except FileNotFoundError as e:\n",
    "        print(e)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "16306a81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there is an issue with my code\n",
      "(1, 2, 3, 4, 5, 6, 7)\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    l1 = [10,20,30,400,10]\n",
    "    l1[100]\n",
    "except:\n",
    "    print(\"there is an issue with my code\")\n",
    "    t = (1,2,3,4,5,6,7)\n",
    "    print(t)\n",
    "    try:\n",
    "        t[0]= 'gogo'\n",
    "        try:\n",
    "            list(t)\n",
    "            print(t)\n",
    "        except:\n",
    "            pass\n",
    "    except:\n",
    "        pass\n",
    "else:\n",
    "    print(\"there is no issue with my code\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "01d08a36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is no issue with this code\n",
      "This is the always executable block\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    l1 = [10,20,30,40,50]\n",
    "except IndexError as e:\n",
    "    print(e)\n",
    "else:\n",
    "    print('There is no issue with this code')\n",
    "finally:\n",
    "    print(\"This is the always executable block\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "55b66fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def askforInt():\n",
    "    while True:\n",
    "        try:\n",
    "            a = int(input(\"Enter a number : \"))\n",
    "            if type(a)==int:\n",
    "                break\n",
    "        except Exception as e:\n",
    "            print(\"this is the error msg\",e)\n",
    "        else:\n",
    "            print(\"person has entered a correct value\")\n",
    "        finally:\n",
    "            print(\"close this issue\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "58a40c3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a number : 10\n",
      "close this issue\n"
     ]
    }
   ],
   "source": [
    "askforInt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "ea9f2aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def askforInt():\n",
    "#     while True:\n",
    "#         try:\n",
    "#             a = float(input(\"Enter a number : \"))\n",
    "#             if type(a)==float:\n",
    "#                 break\n",
    "#         except Exception as e:\n",
    "#             print(\"this is the error msg\",e)\n",
    "#         else:\n",
    "#             print(\"person has entered a correct value\")\n",
    "#         finally:\n",
    "#             print(\"close this issue\")\n",
    "#         finally:\n",
    "#             print(\"finally2\")\n",
    "            \n",
    "# The code provided will raise a SyntaxError because there are two finally blocks.\n",
    "# Python does not allow multiple finally blocks within the same try statement. \n",
    "# A try block can only have one finally block, which is optional but must be unique if present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "dc14c5e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the number : 6\n"
     ]
    },
    {
     "ename": "Exception",
     "evalue": "6",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mException\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[168], line 5\u001b[0m\n\u001b[0;32m      3\u001b[0m a \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28minput\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEnter the number : \u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m a\u001b[38;5;241m==\u001b[39m\u001b[38;5;241m6\u001b[39m:\n\u001b[1;32m----> 5\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m(a)\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m      7\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNot Exception\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mException\u001b[0m: 6"
     ]
    }
   ],
   "source": [
    "# Here in the program we are explicitly raising the exception with the help of raise keyword. \n",
    "# In the program we have mentioned the Exception class Exception\n",
    "\n",
    "# The argument a is passed to the __init__ method of the Exception class. This argument is usually a message or relevant information about the exception.\n",
    "# The created exception object is then \"raised\" and needs to be handled (or the program terminates).\n",
    "\n",
    "a = int(input(\"Enter the number : \"))\n",
    "if a==6:\n",
    "    raise Exception(a)\n",
    "else:\n",
    "    print(\"Not Exception\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "85db486d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the number : 6\n",
      "('the error msg is', 6)\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    a = int(input(\"Enter the number : \"))\n",
    "    if a==6:\n",
    "        raise Exception(\"the error msg is\",a)\n",
    "    else:\n",
    "        print(\"Not Exception\")\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "870afb7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by one\n"
     ]
    }
   ],
   "source": [
    "# here ZeroDivisionError is the subclass of the Exception class \n",
    "# so the instance of the ZeroDvisionError can be assign to the variable e.\n",
    "def create_your_Exception(a,b):\n",
    "    if b==1:\n",
    "        raise ZeroDivisionError('division by one')\n",
    "    else:\n",
    "        print(a/b)\n",
    "\n",
    "try:\n",
    "    create_your_Exception(10,1)\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "b505d147",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n",
      "e is the instance of the Exception class True\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(10/0)\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "    print(\"e is the instance of the Exception class\",isinstance(e,Exception))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c29386",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
